Abstraction Layer for Hybrid KEMs

FAlbertDev opened the pull request:
We plan to add various KEM combiners in the following weeks/months. A KEM combiner is a KEM (with a KEM interface) that internally consists of two (or more) KEMs and/or key exchange algorithms transformed into KEMs. They are used for combining PQC with a classical public key algorithm. For that, this PR defines an abstract interface, a common base for these combiners. 

In general, each KEM combiner consists of multiple public/private keys stored internally that are used to encapsulate multiple shared secrets. These multiple shared secrets are combined (using some sort of KDF) into a single shared secret. For that, the abstraction stores multiple public/private keys and implements the common boilerplate, such as defining the overall strength by returning the strength of the strongest sub-algorithm, etc. Also, a convenient interface for Encryptors and Decryptors is implemented. The existing TLS KEM combiner has been refactored to use this hybrid KEM abstraction.

For the BSI Project 481, we currently plan to implement the following three additional KEM combiners:
- [X-Wing](https://datatracker.ietf.org/doc/draft-connolly-cfrg-xwing-kem/) (Draft): Combines X25519 with ML-KEM-768. Developed by Cloudflare.
- [Ounsworth KEM-Combiner](https://datatracker.ietf.org/doc/draft-ounsworth-cfrg-kem-combiners/) (Draft): Can combine any KEMs. Backed by BSI.
- [CatKDF and/or CasKDF](https://www.etsi.org/deliver/etsi_ts/103700_103799/103744/01.01.01_60/ts_103744v010101p.pdf):  Combines ECDH with any PQC KEM. It is defined as a KEX, but a translation to KEM is given. ETSI standard, recommended in [BSI TR-02102-1 Section B.1.2.](https://www.bsi.bund.de/SharedDocs/Downloads/EN/BSI/Publications/TechGuidelines/TG02102/BSI-TR-02102-1.pdf?__blob=publicationFile&v=7).

FAlbertDev commented on src/lib/pubkey/hybrid_kem/hybrid_kem.h:
```diff
@@ -0,0 +1,121 @@
+ * is applied to derive the shared secret using the individual shared secrets, ciphertexts,
+ * and other context information.
+ */
+class BOTAN_TEST_API Hybrid_PublicKey : public virtual Public_Key {
```
Currently this abstract interface is only in an internal module. I also do not want to commit to a fixed interface (yet). `BOTAN_PUBLIC_API` is only used, if the class itself is exposed, isn't it?

reneme replied:
You're right. Fair enough.

FAlbertDev commented:
Rebased to master

FAlbertDev commented:
Rebased to master.

randombit commented on src/lib/pubkey/hybrid_kem/hybrid_kem.cpp:
```diff
@@ -0,0 +1,88 @@
+   std::transform(
+      public_keys().begin(), public_keys().end(), std::back_inserter(new_private_keys), [&](const auto& public_key) {
+         return public_key->generate_another(rng);
+      });
```
```
for(const auto& pk : public_keys()) {
  new_private_keys.push_back(pk.generate_another(rng));
}
```

reneme commented on src/lib/pubkey/hybrid_kem/hybrid_kem.cpp:
```diff
@@ -0,0 +1,88 @@
+      std::all_of(m_sks.begin(),
+                  m_sks.end(),
+                  [](const auto& sk) { return sk->supports_operation(PublicKeyOperation::KeyEncapsulation); }),
+      "Some provided secret key is not compatible with this hybrid wrapper");
```
What was a holding us back from ranges again? 😅 Was it Xcode/NDK again?

FAlbertDev commented on src/lib/pubkey/hybrid_kem/hybrid_kem.cpp:
```diff
@@ -0,0 +1,88 @@
+      "Some provided public key is not compatible with this hybrid wrapper");
+   m_key_length = reduce(m_pks, size_t(0), [](size_t kl, const auto& key) { return std::max(kl, key->key_length()); });
+   m_estimated_strength =
+      reduce(m_pks, size_t(0), [](size_t es, const auto& key) { return std::max(es, key->estimated_strength()); });
```
With ranges, this would look like
```cpp
m_key_length = std::ranges::max(m_pks | std::views::transform([](const auto& pk) { return pk->key_length(); }));
```
which is more readable than reduce, but in this case, longer and more complex than the simple for-loop. I agree that most parts are easier to understand using only for loops ;)

FAlbertDev commented:
Thanks for the review. I agree that loops are easier to read here. It's quite sad that these C++ iterators are so bulky :(

I addressed your review comments.

FAlbertDev commented:
I also made the KEX to KEM adapter header public. This is very useful for applications using non-predefined KEX algorithms with KEM combiners. Let me know what you think.

FAlbertDev commented:
Rebased to master

FAlbertDev commented on src/lib/pubkey/kex_to_kem_adapter/kex_to_kem_adapter.h:
```diff
@@ -15,13 +15,13 @@
  * Adapter to use a key agreement key pair (e.g. ECDH) as a key encapsulation
  * mechanism.
  */
-class BOTAN_TEST_API KEX_to_KEM_Adapter_PublicKey : public virtual Public_Key {
```
This KEX-to-KEM construction may be quite useful for some applications, such as when using custom protocols that should support KEX and KEM or even when implementing other KEM combiners. 
However, I also see the benefits of hiding this construction in an internal module. I'll make this module internal for now until someone asks for it :P

FAlbertDev commented:
@randombit Sorry for the late update. I applied your review suggestions and cleaned up the history (to make it less bulky in the depending PRs).

randombit reviewed (changes requested):
@FAlbertDev I think it's fine to merge this except we should at least from the outset have `hybrid_kem.h` be an internal only header. We can expose it as a public interface as we gain experience and have more confidence that this is behaving the way we need it to for a variety of cases.

FAlbertDev commented:
Alright! I internalized the header and rebased to the current master 👍

randombit commented:
Didn't see this until just now, pushing to 3.8 but we can merge early on and maybe then consider some of the dependent PRs.

FAlbertDev commented:
Thanks for the approval! Should we wait a few days with this merge in case we need another patch release? 
What do you think?

randombit commented:
@FAlbertDev No go ahead and merge, already `master` is well on its way to a new feature release.
